src/devices/bus/nscsi/tape.cpp

```diff
@@ -0,0 +1,1028 @@
+		case SC_READ_POSITION:                  return handle_read_position();
+		case SC_RELEASE_UNIT:                   return handle_release_unit();
+		case SC_RESERVE_UNIT:                   return handle_reserve_unit();
+		case SC_REWIND:                         return handle_rewind();
+		case SC_SPACE:                          return handle_space();
+		case SC_WRITE_6:                        return handle_write_6();
+		case SC_WRITE_FILEMARKS:                return handle_write_filemarks();
+		// TODO: support more optional commands, once we need to
+		case SC_CHANGE_DEFINITION:
+		case SC_COMPARE:
+		case SC_COPY:
+		case SC_COPY_AND_VERIFY:
+		case SC_LOG_SELECT:
+		case SC_LOG_SENSE:
+		case SC_MODE_SELECT_10:
+		case SC_MODE_SENSE_10:
+		case SC_READ_BUFFER:
+		case SC_READ_REVERSE:
+		case SC_RECEIVE_DIAGNOSTIC_RESULTS:
+		case SC_RECOVER_BUFFERED_DATA:
+		case SC_VERIFY_6:
+		case SC_WRITE_BUFFER:
+		default: // error: unsupported command
+			return report_bad_cmd(cmd);
+	}
+}
+
+uint8_t nscsi_tape_device::scsi_get_data(int id, int pos)
+{
+	switch (id) {
+		// we're handling READ(6); pos = 0 .. (m_rw_req_blocks_num * m_rw_req_block_len - 1)
+		case TAPE_RW_BUF_ID: {
+			const int relative_pos = pos % m_rw_req_block_len;
+			if (m_rw_pending && relative_pos == 0) { // we need to fill buffer by reading another block
+				assert(m_rw_len == 0);
+				continue_handling_read_6();
+			}
+			if (m_rw_len == 0) // buffer is still empty, so reading block failed; condition has already been reported; all we can do here is pretend to read 0 from buffer
+				return 0;
+
+			m_rw_len--;
+			return m_rw_buf[relative_pos];
+		}
+
+		default:
+			return nscsi_full_device::scsi_get_data(id, pos);
+	}
+}
+
+void nscsi_tape_device::scsi_put_data(int id, int pos, uint8_t data)
+{
+	switch (id) {
+		// we're handling WRITE(6); pos == 0 .. (m_rw_req_blocks_num * m_rw_req_block_len - 1)
+		case TAPE_RW_BUF_ID: {
+			if (m_rw_len == m_rw_req_block_len) // buffer is still full, so writing block failed; condition has already been reported; all we can do is pretend to write to buffer
+				return;
+
+			const int relative_pos = pos % m_rw_req_block_len;
+			m_rw_buf[relative_pos] = data;
+			m_rw_len++;
+			if (m_rw_pending && relative_pos == m_rw_req_block_len - 1) { // we need to empty buffer by writing another block
+				assert(m_rw_len == m_rw_req_block_len);
+				continue_handling_write_6();
+			}
+			return;
+		}
+
+		// we're handling MODE SELECT(6); pos == 0 .. (m_pl_len - 1)
+		case TAPE_PL_BUF_ID:
+			m_pl_buf[pos] = data;
+			if (pos == m_pl_len - 1)
+				continue_handling_mode_select_6();
+			return;
+
+		default:
+			return nscsi_full_device::scsi_put_data(id, pos, data);
+	}
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+// condition reporting
+
+void nscsi_tape_device::set_sense_data(const int sense_key, const int sense_key_code, const sense_data data) // TODO: move this function to nscsi_full_device
```

Comment:
I think this (and the report_x() helpers) should be promoted into nscsi_full_device.